src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,212 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public MyLinkedList() {
+
+    }
+
+    public MyLinkedList(int size, Node<T> head, Node<T> tail) {
+        this.size = size;
+        this.head = head;
+        this.tail = tail;
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size);
+        if (index == 0) {
+            addFirst(value);
+        } else if (index == size) {
+            addLast(value);
+        } else {
+            Node<T> oldNode = findElement(index);
+            Node<T> newNode = new Node<T>(oldNode.prev, value, oldNode);
+            oldNode.prev.next = newNode;
+            oldNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        return findElement(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T removed = remove(index);
+        add(value, index);
+        return removed;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        if (index == 0) {
+            return removeFirst();
+        } else if (index == size - 1) {
+            return removeLast();
+        } else {
+            Node<T> oldNode = findElement(index);
+            unlink(oldNode);
```

Comment:
could we implement logic of unlinking node, considering that we could unlink first node or last node and just call unlink method here and in `boolean remove(T object)` method?